MVP ROS 2 <-> gRPC bridge server (#82)
AGH-CEAI/aegis_ros: aegis_grpc/README.md

# aegis_grpc

This packages provides a bridge between ROS 2 and [gRPC](https://grpc.io) protocols, with custom [Protobuf](https://protobuf.dev) messages based on selected ones from [`control_msgs`](https://github.com/ros-controls/control_msgs/tree/humble), [`geometry_msgs`](https://github.com/ros2/common_interfaces/tree/humble/geometry_msgs) and [`sensors_mgs`](https://github.com/ros2/common_interfaces/tree/rolling/sensor_msgs) packages.

It aims to provide a form of "frequency clutch" between real-time domain and low-frequnecy decision making (e.g. an ANN inference in Python).

## Usage

### Launch

``` bash
ros2 launch aegis_grpc start_server.launch.py
# or
ros2 run aegis_grpc grpc_server
```

### Interaction

> The server have enabled a "reflection" plugin, which allows to explore all services with external tool like [`grpcurl`](https://github.com/fullstorydev/grpcurl).

Author: ```suggestion
The server has enabled a "reflection" plugin, which allows you to explore all services with external tools like [`grpcurl`](https://github.com/fullstorydev/grpcurl).
```

```bash
grpcurl -plaintext 127.0.0.1:50051 list
grpcurl -plaintext 127.0.0.1:50051 list proto_aegis_grpc.v1.RobotControlService
grpcurl -plaintext 127.0.0.1:50051 list proto_aegis_grpc.v1.RobotReadService
grpcurl -plaintext 127.0.0.1:50051 describe proto_aegis_grpc.v1.RobotReadService.GetAll
# or with tool from container
podman run --network=host docker.io/fullstorydev/grpcurl -plaintext 127.0.0.1:50051 list
```

> Example call to `GetAll` result as a plain json:

Author: ```suggestion
Example call to the `GetAll` method with result as a plain json:
```

```bash
grpcurl -plaintext 127.0.0.1:50051 \
  -d '{}' \
  127.0.0.1:50051 \
  proto_aegis_grpc.v1.RobotReadService.GetAll
```

## Messages architecture

Since this project aims to bridge communication between ROS and gRPC, the protobuf definitions tries to mimic default messages, services and actions from ROS.
The main difference is the lack of the header with timestamps - the data synchronization responsibility should not be forwarded outside the ROS ecosystem.

The server is split into 2 services defined in [`proto_aegis_grpc.v1.robot_srvs`](./proto_aegis_grpc/v1/robot_srvs.proto): `RobotReadService` and `RobotControlService`.

### Reading data with `RobotReadService`

> The "ROS-getters" are implemented in the [`RobotReadServiceImpl`](./include/aegis_grpc/robot_read_service.hpp) class as the following services:

Author: ```suggestion
The "ROS-getters" are implemented in the [`RobotReadServiceImpl`](./include/aegis_grpc/robot_read_service.hpp) class as the following methods:
```

| ---------------------------------------------------- | --------------------------- | ----- | ----------------------- | --------------------------------------------------------------------------------------- |
| `proto_aegis_grpc.v1.RobotReadService.GetAll`        | Get the all available data. | ✅     | `google.protobuf.Empty` | [`proto_aegis_grpc.v1.robot_srvs.RobotState`](./proto_aegis_grpc/v1/robot_srvs.proto)   |
| `proto_aegis_grpc.v1.RobotReadService.GetJointState` | Get the joints' states.     | ✅     | `google.protobuf.Empty` | [`proto_aegis_grpc.v1.sensor_msgs.JointState`](./proto_aegis_grpc/v1/sensor_msgs.proto) |
| `proto_aegis_grpc.v1.RobotReadService.GetTCPPose`    | Get the TCP pose.           | ✅     | `google.protobuf.Empty` | [`proto_aegis_grpc.v1.geometry_msgs.Pose`](./proto_aegis_grpc/v1/geometry_msgs.proto)   |
| `proto_aegis_grpc.v1.RobotReadService.GetWrench`     | Read force/torque sensor.   | ✅     | `google.protobuf.Empty` | [`proto_aegis_grpc.v1.geometry_msgs.Wrench`](./proto_aegis_grpc/v1/geometry_msgs.proto) |

> You can always list the above services them with the `grpcurl`:

Author: ```suggestion
You can always list the methods with the `grpcurl` command:
```

```bash
grpcurl -plaintext 127.0.0.1:50051 list proto_aegis_grpc.v1.RobotReadService
```

### Controlling the robot with `RobotControlService`

> The control bridge for MoveIt2 servo and planners are implemented in the [`RobotControlServiceImpl`](./include/aegis_grpc/robot_control_service.hpp) class as the following services:

Author: ```suggestion
The control bridge for MoveIt2 servo and planners are implemented in the [`RobotControlServiceImpl`](./include/aegis_grpc/robot_control_service.hpp) class as the following methods:
```

| `proto_aegis_grpc.v1.RobotControlService.GriperSetWidth` | Set the gripper width.                                          | ✅     | [`proto_aegis_grpc.v1.robot_srvs.GripperSetWidthRequest`](./proto_aegis_grpc/v1/robot_srvs.proto) | [`proto_aegis_grpc.v1.robot_srvs.TriggerResponse`](./proto_aegis_grpc/v1/robot_srvs.proto) |
| `proto_aegis_grpc.v1.RobotControlService.ServoJoint`     | Servoing (w.r.t. joints).                                       | ✅     | [`proto_aegis_grpc.v1.control_msgs.JointJog`](./proto_aegis_grpc/v1/control_msgs.proto)           | `google.protobuf.Empty`                                                                    |
| `proto_aegis_grpc.v1.RobotControlService.ServoTCP`       | Servoing (w.r.t. TCP).                                          | ✅     | [`proto_aegis_grpc.v1.geometry_msgs.Twist`](./proto_aegis_grpc/v1/geometry_msgs.proto)            | `google.protobuf.Empty`                                                                    |

> You can always list the above services them with the `grpcurl`:

Author: ```suggestion
You can always list the methods with the `grpcurl` command:
```

| `proto_aegis_grpc.v1.RobotControlService.GotoJoints`         | Plan & execute the MoveIt2 trajectory to a given joints target. | ❌     | [`proto_aegis_grpc.v1.sensor_msgs.JointState`](./proto_aegis_grpc/v1/sensor_msgs.proto)              | [`proto_aegis_grpc.v1.robot_srvs.TriggerResponse`](./proto_aegis_grpc/v1/robot_srvs.proto) |
| `proto_aegis_grpc.v1.RobotControlService.GotoPose`           | Plan & execute the MoveIt2 trajectory to a given pose target.   | ❌     | [`proto_aegis_grpc.v1.geometry_msgs.Pose`](./proto_aegis_grpc/v1/geometry_msgs.proto)                | [`proto_aegis_grpc.v1.robot_srvs.TriggerResponse`](./proto_aegis_grpc/v1/robot_srvs.proto) |
| `proto_aegis_grpc.v1.RobotControlService.GriperClose`        | Closes the gripper.                                             | ✅     | `google.protobuf.Empty`                                                                              | [`proto_aegis_grpc.v1.robot_srvs.TriggerResponse`](./proto_aegis_grpc/v1/robot_srvs.proto) |

> | `proto_aegis_grpc.v1.RobotControlService.GriperOpen`         | Opens the gripper.                                              | ✅     | `google.protobuf.Empty`                                                                              | [`proto_aegis_grpc.v1.robot_srvs.TriggerResponse`](./proto_aegis_grpc/v1/robot_srvs.proto) |

Reviewer: ```suggestion
| `proto_aegis_grpc.v1.RobotControlService.GripperClose`        | Closes the gripper.                                             | ✅     | `google.protobuf.Empty`                                                                              | [`proto_aegis_grpc.v1.robot_srvs.TriggerResponse`](./proto_aegis_grpc/v1/robot_srvs.proto) |
| `proto_aegis_grpc.v1.RobotControlService.GripperOpen`         | Opens the gripper.                                              | ✅     | `google.protobuf.Empty`                                                                              | [`proto_aegis_grpc.v1.robot_srvs.TriggerResponse`](./proto_aegis_grpc/v1/robot_srvs.proto) |
```
